evm/src/lookup.rs

```diff
@@ -1,127 +1,248 @@
+    // for every column, but that would require building a big vector of all the columns concatenated.
+    // Not sure which approach is better.
+    // Note: these are the h_k(x) polynomials in the paper, with a few differences:
+    //       * Here, the first ratio m_0(x)/phi_0(x) is not included with the columns batched up to create the 
+    //         h_k polynomials; instead there's a separate helper column for it (see below).
+    //       * Here, we use 1 instead of -1 as the numerator (and subtract later).
+    //       * Here, for now, the batch size (l) is always constraint_degree - 1 = 2.
+    for mut col_inds in &lookup.columns.iter().chunks(constraint_degree - 1) {
+        let first = *col_inds.next().unwrap();
+        // TODO: The clone could probably be avoided by using a modified version of `batch_multiplicative_inverse`
+        // taking `challenge` as an additional argument.
+        let mut column = trace_poly_values[first].values.clone();
+        for x in column.iter_mut() {
+            *x = challenge + *x;
+        }
+        let mut acc = F::batch_multiplicative_inverse(&column);
+        for &ind in col_inds {
+            let mut column = trace_poly_values[ind].values.clone();
+            for x in column.iter_mut() {
+                *x = challenge + *x;
+            }
+            column = F::batch_multiplicative_inverse(&column);
+            batch_add_inplace(&mut acc, &column);
+        }
+        helper_columns.push(acc.into());
+    }
 
-    // The permuted inputs do not have to be ordered, but we found that sorting was faster than
-    // hash-based grouping. We also sort the table, as this helps us identify "unused" table
-    // elements efficiently.
+    // Add `1/(table+challenge)` to the helper columns.
+    // This is 1/phi_0(x) = 1/(x + t(x)) from the paper.
+    // Here, we don't include m(x) in the numerator, instead multiplying it with this column later.
+    let mut table = trace_poly_values[lookup.table_column].values.clone();
+    for x in table.iter_mut() {
+        *x = challenge + *x;
+    }
+    helper_columns.push(F::batch_multiplicative_inverse(&table).into());
+
+    // Compute the `Z` polynomial with `Z(1)=0` and `Z(gx) = Z(x) + sum h_i(x) - frequencies(x)g(x)`.
+    // This enforces the check from the paper, that the sum of the h_k(x) polynomials is 0 over H.
+    // In the paper, that sum includes m(x)/(x + t(x)) = frequencies(x)/g(x), because that was bundled
+    // into the h_k(x) polynomials.
+    let frequencies = &trace_poly_values[lookup.frequencies_column].values;
+    let mut z = Vec::with_capacity(frequencies.len());
+    z.push(F::ZERO);
+    for i in 0..frequencies.len() - 1 {
+        let x = helper_columns[..num_helper_columns - 2]
+            .iter()
+            .map(|col| col.values[i])
+            .sum::<F>()
+            - frequencies[i] * helper_columns[num_helper_columns - 2].values[i];
+        z.push(z[i] + x);
```

Comment:
Maybe a sanity check, that the last value for `z` (which is not computed here) ends up with `F::ZERO`, would be a nice feature?